File: libs/ts-types/src/settings.types.ts
/**
 * Settings Types
 *
 * Comprehensive type definitions for settings and parental controls.
 * Supports COPPA/GDPR compliance, WCAG 2.1 AA accessibility, and
 * cross-device synchronization.
 *
 * @packageDocumentation
 */

// ============================================================================
// PARENTAL CONTROLS
// ============================================================================

/**
 * Screen time management settings
 * Controls daily limits, schedules, and break reminders
 */
export interface ScreenTimeSettings {
  /** Daily screen time limit in minutes */
  dailyLimit: number;
  /** Whether scheduled access is enabled */
  scheduleEnabled: boolean;
  /** Daily schedule by day of week */
  schedule: Record<
    'monday' | 'tuesday' | 'wednesday' | 'thursday' | 'friday' | 'saturday' | 'sunday',
    DailySchedule
  >;
  /** Whether to show break reminders */
  breakReminders: boolean;
  /** Minutes between break reminders */
  breakInterval: number;
  /** Break duration in minutes */
  breakDuration: number;
}

/**
 * Daily schedule for a single day
 */
export interface DailySchedule {
  /** Whether this day's schedule is enabled */
  enabled: boolean;
  /** Start time in HH:MM format */
  start: string;
  /** End time in HH:MM format */
  end: string;
}

/**
 * Content filtering settings
 * Controls what content is accessible to the learner
 */
export interface ContentFilterSettings {
  /** Content restriction level */
  restrictionLevel: 'strict' | 'moderate' | 'minimal';
  /** Block external links in content */
  blockExternalLinks: boolean;
  /** Enable safe search for any search functionality */
  safeSearch: boolean;
  /** Hide achievement leaderboards (for privacy) */
  hideAchievementLeaderboards: boolean;
  /** Restrict chat/messaging features */
  restrictChat: boolean;
  /** Topics explicitly blocked */
  blockedTopics: string[];
  /** Topics explicitly allowed (overrides restrictions) */
  allowedTopics: string[];
}

/**
 * Subject access controls
 * Controls which subjects and levels are accessible
 */
export interface SubjectAccessSettings {
  /** Per-subject access configuration */
  subjects: Record<string, SubjectConfig>;
  /** Require parent approval before accessing new subjects */
  requireParentApprovalForNewSubjects: boolean;
  /** Allow access to advanced/accelerated content */
  allowAdvancedContent: boolean;
}

/**
 * Configuration for a single subject
 */
export interface SubjectConfig {
  /** Whether the subject is enabled */
  enabled: boolean;
  /** Maximum level allowed (optional cap) */
  maxLevel?: number;
}

/**
 * Privacy and safety settings
 */
export interface SafetySettings {
  /** Privacy visibility settings */
  privacy: PrivacySettings;
  /** Data collection preferences (GDPR/COPPA) */
  data: DataPreferences;
  /** Access control settings */
  access: AccessControlSettings;
  /** Emergency contact settings */
  emergency: EmergencySettings;
}

/**
 * Privacy visibility settings
 */
export interface PrivacySettings {
  /** Hide learner from classmate lists */
  hideFromClassmates: boolean;
  /** Hide profile photo */
  hideProfilePhoto: boolean;
  /** Hide progress from other students */
  hideProgress: boolean;
  /** Enable anonymous mode for social features */
  anonymousMode: boolean;
}

/**
 * Data collection preferences (COPPA/GDPR compliant)
 */
export interface DataPreferences {
  /** Allow analytics collection */
  allowAnalytics: boolean;
  /** Allow AI personalization */
  allowPersonalization: boolean;
  /** Allow third-party data sharing */
  allowThirdPartySharing: boolean;
}

/**
 * Access control settings
 */
export interface AccessControlSettings {
  /** Require PIN for changing parental control settings */
  requirePinForSettings: boolean;
  /** Hashed PIN (never store plaintext) */
  pinHash?: string;
  /** Salt for PIN hashing */
  pinSalt?: string;
  /** Allow account deletion by learner */
  allowAccountDeletion: boolean;
  /** Two-factor authentication enabled */
  twoFactorEnabled: boolean;
}

/**
 * Emergency contact settings
 */
export interface EmergencySettings {
  /** Emergency contact feature enabled */
  emergencyContactEnabled: boolean;
  /** Emergency contact email */
  emergencyEmail?: string;
  /** Emergency contact phone */
  emergencyPhone?: string;
}

/**
 * Complete parental controls configuration
 */
export interface ParentalControlSettings {
  /** Screen time management */
  screenTime: ScreenTimeSettings;
  /** Content filtering */
  contentFilter: ContentFilterSettings;
  /** Subject access controls */
  subjectAccess: SubjectAccessSettings;
  /** Notification preferences */
  notifications: NotificationSettings;
  /** Safety and privacy settings */
  safety: SafetySettings;
}

// ============================================================================
// NOTIFICATION SETTINGS
// ============================================================================

/**
 * Email notification preferences
 */
export interface EmailNotificationSettings {
  /** Master email toggle */
  enabled: boolean;
  /** Daily digest email */
  dailyDigest: boolean;
  /** Weekly report email */
  weeklyReport: boolean;
  /** Achievement notifications */
  achievements: boolean;
  /** Concern/issue notifications */
  concerns: boolean;
  /** Teacher message notifications */
  teacherMessages: boolean;
}

/**
 * Push notification preferences
 */
export interface PushNotificationSettings {
  /** Master push toggle */
  enabled: boolean;
  /** Achievement notifications */
  achievements: boolean;
  /** Milestone notifications */
  milestones: boolean;
  /** Concern notifications */
  concerns: boolean;
  /** Teacher message notifications */
  teacherMessages: boolean;
  /** Screen time limit alerts */
  screenTimeAlerts: boolean;
}

/**
 * In-app notification preferences
 */
export interface InAppNotificationSettings {
  /** Master in-app toggle */
  enabled: boolean;
  /** Achievement notifications */
  achievements: boolean;
  /** Progress notifications */
  progress: boolean;
  /** Recommendation notifications */
  recommendations: boolean;
}

/**
 * Quiet hours configuration
 */
export interface QuietHoursSettings {
  /** Quiet hours enabled */
  enabled: boolean;
  /** Start time in HH:MM format */
  start: string;
  /** End time in HH:MM format */
  end: string;
}

/**
 * Complete notification settings
 */
export interface NotificationSettings {
  /** Email notification preferences */
  email: EmailNotificationSettings;
  /** Push notification preferences */
  push: PushNotificationSettings;
  /** In-app notification preferences */
  inApp: InAppNotificationSettings;
  /** Quiet hours configuration */
  quietHours: QuietHoursSettings;
}

// ============================================================================
// ACCESSIBILITY SETTINGS
// ============================================================================

/**
 * Color blind mode options
 */
export type ColorBlindMode =
  | 'none'
  | 'protanopia' // Red-blind
  | 'deuteranopia' // Green-blind
  | 'tritanopia' // Blue-blind
  | 'achromatopsia'; // Total color blindness

/**
 * Text size presets
 */
export type TextSizePreset = 'small' | 'default' | 'large' | 'extra-large';

/**
 * Accessibility settings for visual, motor, and sensory accommodations
 * Compliant with WCAG 2.1 AA standards
 */
export interface AccessibilitySettings {
  // Visual accommodations
  /** Text scale factor (1.0 = 100%, 1.5 = 150%, etc.) */
  textScaleFactor: number;
  /** Enable high contrast mode */
  highContrast: boolean;
  /** Enable bold text throughout */
  boldText: boolean;
  /** Color blind mode selection */
  colorBlindMode: ColorBlindMode;
  /** Reduce transparency/blur effects */
  reduceTransparency: boolean;

  // Motion accommodations
  /** Reduce motion/animations (prefers-reduced-motion) */
  reduceMotion: boolean;

  // Screen reader support
  /** Optimize UI for screen readers */
  screenReaderOptimized: boolean;

  // Audio/sensory accommodations
  /** Enable haptic feedback */
  hapticFeedbackEnabled: boolean;
  /** Enable sound effects */
  soundEffectsEnabled: boolean;
  /** Sound volume (0.0 - 1.0) */
  soundVolume: number;
  /** Enable read-aloud feature */
  readAloudEnabled: boolean;
  /** Read-aloud speed (0.5 - 2.0) */
  readAloudSpeed: number;
}

// ============================================================================
// APP SETTINGS
// ============================================================================

/**
 * Theme mode options
 */
export type ThemeMode = 'light' | 'dark' | 'system';

/**
 * Supported locales
 */
export type Locale = 'en' | 'es' | 'fr' | 'de' | 'pt' | 'zh' | 'ja' | 'ko' | 'ar' | 'hi';

/**
 * General app settings
 */
export interface AppSettings {
  /** User's preferred locale */
  locale: Locale;
  /** Theme mode preference */
  themeMode: ThemeMode;
  /** Timezone (IANA format, e.g., 'America/New_York') */
  timezone: string;
}

// ============================================================================
// PARENT SETTINGS (COMBINED)
// ============================================================================

/**
 * Complete parent settings including profile and preferences
 */
export interface ParentSettings {
  /** Parent profile information */
  profile: ParentProfileSettings;
  /** General app settings */
  app: AppSettings;
  /** Notification preferences */
  notifications: NotificationSettings;
}

/**
 * Parent profile settings
 */
export interface ParentProfileSettings {
  firstName: string;
  lastName: string;
  email: string;
  phone?: string;
  avatarUrl?: string;
}

// ============================================================================
// LEARNER SETTINGS (COMBINED)
// ============================================================================

/**
 * Complete learner settings
 */
export interface LearnerSettings {
  /** General app settings */
  app: AppSettings;
  /** Accessibility settings */
  accessibility: AccessibilitySettings;
  /** Learning preferences */
  learning: LearningPreferences;
}

/**
 * Learning preference settings
 */
export interface LearningPreferences {
  /** Preferred learning style */
  preferredStyle: 'visual' | 'auditory' | 'kinesthetic' | 'reading';
  /** Daily goal in minutes */
  dailyGoalMinutes: number;
  /** Show hints automatically */
  autoShowHints: boolean;
  /** Preferred difficulty */
  preferredDifficulty: 'easier' | 'normal' | 'challenging';
}

// ============================================================================
// API TYPES
// ============================================================================

/**
 * Request to update parental controls
 */
export interface UpdateParentalControlsRequest<
  K extends keyof ParentalControlSettings = keyof ParentalControlSettings,
> {
  /** Child/student ID */
  childId: string;
  /** Section being updated */
  section: K;
  /** New settings for that section */
  settings: ParentalControlSettings[K];
}

/**
 * Request to verify PIN
 */
export interface VerifyPinRequest {
  /** PIN to verify */
  pin: string;
  /** Action being performed (for audit) */
  action: string;
}

/**
 * Response from PIN verification
 */
export interface VerifyPinResponse {
  /** Whether PIN is correct */
  valid: boolean;
  /** Temporary token for subsequent protected operations */
  token?: string;
  /** Token expiry time */
  expiresAt?: string;
  /** Remaining attempts before lockout */
  remainingAttempts?: number;
}

/**
 * Request to set/update PIN
 */
export interface SetPinRequest {
  /** New PIN (4-6 digits) */
  newPin: string;
  /** Current PIN (required if updating) */
  currentPin?: string;
}

/**
 * Request to update accessibility settings
 */
export interface UpdateAccessibilityRequest {
  settings: Partial<AccessibilitySettings>;
}

/**
 * Settings sync metadata
 */
export interface SettingsSyncMetadata {
  /** Last sync timestamp */
  lastSyncedAt: string;
  /** Device that last modified */
  lastModifiedBy: string;
  /** Settings version for conflict resolution */
  version: number;
}

// ============================================================================
// DEFAULT VALUES
// ============================================================================

/**
 * Default screen time settings
 */
export const DEFAULT_SCREEN_TIME_SETTINGS: ScreenTimeSettings = {
  dailyLimit: 120, // 2 hours
  scheduleEnabled: false,
  schedule: {
    monday: { enabled: true, start: '08:00', end: '20:00' },
    tuesday: { enabled: true, start: '08:00', end: '20:00' },
    wednesday: { enabled: true, start: '08:00', end: '20:00' },
    thursday: { enabled: true, start: '08:00', end: '20:00' },
    friday: { enabled: true, start: '08:00', end: '20:00' },
    saturday: { enabled: true, start: '09:00', end: '21:00' },
    sunday: { enabled: true, start: '09:00', end: '21:00' },
  },
  breakReminders: true,
  breakInterval: 30,
  breakDuration: 5,
};

/**
 * Default content filter settings
 */
export const DEFAULT_CONTENT_FILTER_SETTINGS: ContentFilterSettings = {
  restrictionLevel: 'moderate',
  blockExternalLinks: true,
  safeSearch: true,
  hideAchievementLeaderboards: false,
  restrictChat: true,
  blockedTopics: [],
  allowedTopics: [],
};

/**
 * Default accessibility settings
 */
export const DEFAULT_ACCESSIBILITY_SETTINGS: AccessibilitySettings = {
  textScaleFactor: 1.0,
  highContrast: false,
  boldText: false,
  colorBlindMode: 'none',
  reduceTransparency: false,
  reduceMotion: false,
  screenReaderOptimized: false,
  hapticFeedbackEnabled: true,
  soundEffectsEnabled: true,
  soundVolume: 0.7,
  readAloudEnabled: false,
  readAloudSpeed: 1.0,
};

/**
 * Default notification settings
 */
export const DEFAULT_NOTIFICATION_SETTINGS: NotificationSettings = {
  email: {
    enabled: true,
    dailyDigest: false,
    weeklyReport: true,
    achievements: true,
    concerns: true,
    teacherMessages: true,
  },
  push: {
    enabled: true,
    achievements: true,
    milestones: true,
    concerns: true,
    teacherMessages: true,
    screenTimeAlerts: true,
  },
  inApp: {
    enabled: true,
    achievements: true,
    progress: true,
    recommendations: true,
  },
  quietHours: {
    enabled: false,
    start: '21:00',
    end: '07:00',
  },
};

File: apps/mobile-parent/lib/features/settings/api/settings_types.dart
/// Settings Types for Mobile Parent App
///
/// Type definitions for settings and parental controls.
/// Mirrors the TypeScript types from @aivo/ts-types.

import 'package:freezed_annotation/freezed_annotation.dart';

part 'settings_types.freezed.dart';
part 'settings_types.g.dart';

// ============================================================================
// SCREEN TIME SETTINGS
// ============================================================================

@freezed
class DailySchedule with _$DailySchedule {
  const factory DailySchedule({
    required bool enabled,
    required String start,
    required String end,
  }) = _DailySchedule;

  factory DailySchedule.fromJson(Map<String, dynamic> json) =>
      _$DailyScheduleFromJson(json);
}

@freezed
class ScreenTimeSettings with _$ScreenTimeSettings {
  const factory ScreenTimeSettings({
    required int dailyLimit,
    required bool scheduleEnabled,
    required Map<String, DailySchedule> schedule,
    required bool breakReminders,
    required int breakInterval,
    required int breakDuration,
  }) = _ScreenTimeSettings;

  factory ScreenTimeSettings.fromJson(Map<String, dynamic> json) =>
      _$ScreenTimeSettingsFromJson(json);

  factory ScreenTimeSettings.defaults() => ScreenTimeSettings(
        dailyLimit: 120,
        scheduleEnabled: false,
        schedule: {
          'monday': const DailySchedule(enabled: true, start: '08:00', end: '20:00'),
          'tuesday': const DailySchedule(enabled: true, start: '08:00', end: '20:00'),
          'wednesday': const DailySchedule(enabled: true, start: '08:00', end: '20:00'),
          'thursday': const DailySchedule(enabled: true, start: '08:00', end: '20:00'),
          'friday': const DailySchedule(enabled: true, start: '08:00', end: '20:00'),
          'saturday': const DailySchedule(enabled: true, start: '09:00', end: '21:00'),
          'sunday': const DailySchedule(enabled: true, start: '09:00', end: '21:00'),
        },
        breakReminders: true,
        breakInterval: 30,
        breakDuration: 5,
      );
}

// ============================================================================
// CONTENT FILTER SETTINGS
// ============================================================================

enum RestrictionLevel {
  @JsonValue('strict')
  strict,
  @JsonValue('moderate')
  moderate,
  @JsonValue('minimal')
  minimal,
}

@freezed
class ContentFilterSettings with _$ContentFilterSettings {
  const factory ContentFilterSettings({
    required RestrictionLevel restrictionLevel,
    required bool blockExternalLinks,
    required bool safeSearch,
    required bool hideAchievementLeaderboards,
    required bool restrictChat,
    required List<String> blockedTopics,
    required List<String> allowedTopics,
  }) = _ContentFilterSettings;

  factory ContentFilterSettings.fromJson(Map<String, dynamic> json) =>
      _$ContentFilterSettingsFromJson(json);

  factory ContentFilterSettings.defaults() => const ContentFilterSettings(
        restrictionLevel: RestrictionLevel.moderate,
        blockExternalLinks: true,
        safeSearch: true,
        hideAchievementLeaderboards: false,
        restrictChat: true,
        blockedTopics: [],
        allowedTopics: [],
      );
}

// ============================================================================
// SUBJECT ACCESS SETTINGS
// ============================================================================

@freezed
class SubjectConfig with _$SubjectConfig {
  const factory SubjectConfig({
    required bool enabled,
    int? maxLevel,
  }) = _SubjectConfig;

  factory SubjectConfig.fromJson(Map<String, dynamic> json) =>
      _$SubjectConfigFromJson(json);
}

@freezed
class SubjectAccessSettings with _$SubjectAccessSettings {
  const factory SubjectAccessSettings({
    required Map<String, SubjectConfig> subjects,
    required bool requireParentApprovalForNewSubjects,
    required bool allowAdvancedContent,
  }) = _SubjectAccessSettings;

  factory SubjectAccessSettings.fromJson(Map<String, dynamic> json) =>
      _$SubjectAccessSettingsFromJson(json);

  factory SubjectAccessSettings.defaults() => const SubjectAccessSettings(
        subjects: {},
        requireParentApprovalForNewSubjects: true,
        allowAdvancedContent: false,
      );
}

// ============================================================================
// SAFETY SETTINGS
// ============================================================================

@freezed
class PrivacySettings with _$PrivacySettings {
  const factory PrivacySettings({
    required bool hideFromClassmates,
    required bool hideProfilePhoto,
    required bool hideProgress,
    required bool anonymousMode,
  }) = _PrivacySettings;

  factory PrivacySettings.fromJson(Map<String, dynamic> json) =>
      _$PrivacySettingsFromJson(json);
}

@freezed
class DataPreferences with _$DataPreferences {
  const factory DataPreferences({
    required bool allowAnalytics,
    required bool allowPersonalization,
    required bool allowThirdPartySharing,
  }) = _DataPreferences;

  factory DataPreferences.fromJson(Map<String, dynamic> json) =>
      _$DataPreferencesFromJson(json);
}

@freezed
class AccessControlSettings with _$AccessControlSettings {
  const factory AccessControlSettings({
    required bool requirePinForSettings,
    String? pinHash,
    String? pinSalt,
    required bool allowAccountDeletion,
    required bool twoFactorEnabled,
  }) = _AccessControlSettings;

  factory AccessControlSettings.fromJson(Map<String, dynamic> json) =>
      _$AccessControlSettingsFromJson(json);
}

@freezed
class EmergencySettings with _$EmergencySettings {
  const factory EmergencySettings({
    required bool emergencyContactEnabled,
    String? emergencyEmail,
    String? emergencyPhone,
  }) = _EmergencySettings;

  factory EmergencySettings.fromJson(Map<String, dynamic> json) =>
      _$EmergencySettingsFromJson(json);
}

@freezed
class SafetySettings with _$SafetySettings {
  const factory SafetySettings({
    required PrivacySettings privacy,
    required DataPreferences data,
    required AccessControlSettings access,
    required EmergencySettings emergency,
  }) = _SafetySettings;

  factory SafetySettings.fromJson(Map<String, dynamic> json) =>
      _$SafetySettingsFromJson(json);
}

// ============================================================================
// NOTIFICATION SETTINGS
// ============================================================================

@freezed
class EmailNotificationSettings with _$EmailNotificationSettings {
  const factory EmailNotificationSettings({
    required bool enabled,
    required bool dailyDigest,
    required bool weeklyReport,
    required bool achievements,
    required bool concerns,
    required bool teacherMessages,
  }) = _EmailNotificationSettings;

  factory EmailNotificationSettings.fromJson(Map<String, dynamic> json) =>
      _$EmailNotificationSettingsFromJson(json);
}

@freezed
class PushNotificationSettings with _$PushNotificationSettings {
  const factory PushNotificationSettings({
    required bool enabled,
    required bool achievements,
    required bool milestones,
    required bool concerns,
    required bool teacherMessages,
    required bool screenTimeAlerts,
  }) = _PushNotificationSettings;

  factory PushNotificationSettings.fromJson(Map<String, dynamic> json) =>
      _$PushNotificationSettingsFromJson(json);
}

@freezed
class InAppNotificationSettings with _$InAppNotificationSettings {
  const factory InAppNotificationSettings({
    required bool enabled,
    required bool achievements,
    required bool progress,
    required bool recommendations,
  }) = _InAppNotificationSettings;

  factory InAppNotificationSettings.fromJson(Map<String, dynamic> json) =>
      _$InAppNotificationSettingsFromJson(json);
}

@freezed
class QuietHoursSettings with _$QuietHoursSettings {
  const factory QuietHoursSettings({
    required bool enabled,
    required String start,
    required String end,
  }) = _QuietHoursSettings;

  factory QuietHoursSettings.fromJson(Map<String, dynamic> json) =>
      _$QuietHoursSettingsFromJson(json);
}

@freezed
class NotificationSettings with _$NotificationSettings {
  const factory NotificationSettings({
    required EmailNotificationSettings email,
    required PushNotificationSettings push,
    required InAppNotificationSettings inApp,
    required QuietHoursSettings quietHours,
  }) = _NotificationSettings;

  factory NotificationSettings.fromJson(Map<String, dynamic> json) =>
      _$NotificationSettingsFromJson(json);

  factory NotificationSettings.defaults() => const NotificationSettings(
        email: EmailNotificationSettings(
          enabled: true,
          dailyDigest: false,
          weeklyReport: true,
          achievements: true,
          concerns: true,
          teacherMessages: true,
        ),
        push: PushNotificationSettings(
          enabled: true,
          achievements: true,
          milestones: true,
          concerns: true,
          teacherMessages: true,
          screenTimeAlerts: true,
        ),
        inApp: InAppNotificationSettings(
          enabled: true,
          achievements: true,
          progress: true,
          recommendations: true,
        ),
        quietHours: QuietHoursSettings(
          enabled: false,
          start: '21:00',
          end: '07:00',
        ),
      );
}

// ============================================================================
// PARENTAL CONTROL SETTINGS (COMBINED)
// ============================================================================

@freezed
class ParentalControlSettings with _$ParentalControlSettings {
  const factory ParentalControlSettings({
    required ScreenTimeSettings screenTime,
    required ContentFilterSettings contentFilter,
    required SubjectAccessSettings subjectAccess,
    required NotificationSettings notifications,
    required SafetySettings safety,
  }) = _ParentalControlSettings;

  factory ParentalControlSettings.fromJson(Map<String, dynamic> json) =>
      _$ParentalControlSettingsFromJson(json);
}

// ============================================================================
// APP SETTINGS
// ============================================================================

enum ThemeMode {
  @JsonValue('light')
  light,
  @JsonValue('dark')
  dark,
  @JsonValue('system')
  system,
}

@freezed
class AppSettings with _$AppSettings {
  const factory AppSettings({
    required String locale,
    required ThemeMode themeMode,
    required String timezone,
  }) = _AppSettings;

  factory AppSettings.fromJson(Map<String, dynamic> json) =>
      _$AppSettingsFromJson(json);
}

// ============================================================================
// PARENT PROFILE SETTINGS
// ============================================================================

@freezed
class ParentProfileSettings with _$ParentProfileSettings {
  const factory ParentProfileSettings({
    required String firstName,
    required String lastName,
    required String email,
    String? phone,
    String? avatarUrl,
  }) = _ParentProfileSettings;

  factory ParentProfileSettings.fromJson(Map<String, dynamic> json) =>
      _$ParentProfileSettingsFromJson(json);
}

// ============================================================================
// PARENT SETTINGS (COMBINED)
// ============================================================================

@freezed
class ParentSettings with _$ParentSettings {
  const factory ParentSettings({
    required ParentProfileSettings profile,
    required AppSettings app,
    required NotificationSettings notifications,
  }) = _ParentSettings;

  factory ParentSettings.fromJson(Map<String, dynamic> json) =>
      _$ParentSettingsFromJson(json);
}

// ============================================================================
// PIN PROTECTION
// ============================================================================

@freezed
class PinStatus with _$PinStatus {
  const factory PinStatus({
    required bool enabled,
  }) = _PinStatus;

  factory PinStatus.fromJson(Map<String, dynamic> json) =>
      _$PinStatusFromJson(json);
}

@freezed
class VerifyPinResponse with _$VerifyPinResponse {
  const factory VerifyPinResponse({
    required bool valid,
    String? token,
    String? expiresAt,
    int? remainingAttempts,
  }) = _VerifyPinResponse;

  factory VerifyPinResponse.fromJson(Map<String, dynamic> json) =>
      _$VerifyPinResponseFromJson(json);
}

// ============================================================================
// SETTINGS SYNC
// ============================================================================

@freezed
class SettingsSyncMetadata with _$SettingsSyncMetadata {
  const factory SettingsSyncMetadata({
    required String lastSyncedAt,
    required String lastModifiedBy,
    required int version,
  }) = _SettingsSyncMetadata;

  factory SettingsSyncMetadata.fromJson(Map<String, dynamic> json) =>
      _$SettingsSyncMetadataFromJson(json);
}

// ============================================================================
// ACCESSIBILITY SETTINGS
// ============================================================================

enum ColorBlindMode {
  @JsonValue('none')
  none,
  @JsonValue('protanopia')
  protanopia,
  @JsonValue('deuteranopia')
  deuteranopia,
  @JsonValue('tritanopia')
  tritanopia,
  @JsonValue('achromatopsia')
  achromatopsia,
}

@freezed
class AccessibilitySettings with _$AccessibilitySettings {
  const factory AccessibilitySettings({
    required double textScaleFactor,
    required bool highContrast,
    required bool boldText,
    required ColorBlindMode colorBlindMode,
    required bool reduceTransparency,
    required bool reduceMotion,
    required bool screenReaderOptimized,
    required bool hapticFeedbackEnabled,
    required bool soundEffectsEnabled,
    required double soundVolume,
    required bool readAloudEnabled,
    required double readAloudSpeed,
  }) = _AccessibilitySettings;

  factory AccessibilitySettings.fromJson(Map<String, dynamic> json) =>
      _$AccessibilitySettingsFromJson(json);

  factory AccessibilitySettings.defaults() => const AccessibilitySettings(
        textScaleFactor: 1.0,
        highContrast: false,
        boldText: false,
        colorBlindMode: ColorBlindMode.none,
        reduceTransparency: false,
        reduceMotion: false,
        screenReaderOptimized: false,
        hapticFeedbackEnabled: true,
        soundEffectsEnabled: true,
        soundVolume: 0.7,
        readAloudEnabled: false,
        readAloudSpeed: 1.0,
      );
}
